test/jdk/java/lang/String/StringRacyConstructor.java

```diff
@@ -0,0 +1,418 @@
+        assertEquals(orig, stringFromChars, "mixed chars");
+        assertTrue(validCoder(stringFromChars), "invalid coder"
+                + ", invalid coder: " + inspectString(stringFromChars));
+    }
+
+    private static List<String> strings() {
+        return List.of("01234", " ");
+    }
+
+    @ParameterizedTest
+    @MethodSource("strings")
+    public void racyString(String orig) {
+        String racyString = racyStringConstruction(orig);
+        // The contents are indeterminate due to the race
+        assertTrue(validCoder(racyString), orig + " string invalid"
+                + ", racyString: " + inspectString(racyString));
+    }
+
+    @ParameterizedTest
+    @MethodSource("strings")
+    public void racyCodePoint(String orig) {
+        String iffyString = racyStringConstructionCodepoints(orig);
+        // The contents are indeterminate due to the race
+        assertTrue(validCoder(iffyString), "invalid coder in non-deterministic string"
+                + ", orig:" + inspectString(orig)
+                + ", iffyString: " + inspectString(iffyString));
+    }
+
+    @ParameterizedTest
+    @MethodSource("strings")
+    public void racyCodePointSurrogates(String orig) {
+        String iffyString = racyStringConstructionCodepointsSurrogates(orig);
+        // The contents are indeterminate due to the race
+        if (!orig.equals(iffyString))
+            System.err.println("orig: " + orig + ", iffy: " + iffyString + Arrays.toString(iffyString.codePoints().toArray()));
+        assertTrue(validCoder(iffyString), "invalid coder in non-deterministic string"
+                + ", orig:" + inspectString(orig)
+                + ", iffyString: " + inspectString(iffyString));
+    }
+
+    // Test the private methods of StringUTF16 that compress and copy COMPRESSED_STRING
+    // encoded byte arrays.
+    @Test
+    public void verifyUTF16CopyBytes()
+            throws ClassNotFoundException, NoSuchMethodException, InvocationTargetException, IllegalAccessException {
+        Class<?> stringUTF16 = Class.forName("java.lang.StringUTF16");
+        Method mCompressChars = stringUTF16.getDeclaredMethod("compress",
+                char[].class, int.class, byte[].class, int.class, int.class);
+        mCompressChars.setAccessible(true);
+
+        // First warmup the intrinsic and check 1 case
+        char[] chars = {'a', 'b', 'c', 0xff21, 0xff22, 0xff23};
+        byte[] bytes = new byte[chars.length];
+        int printWarningCount = 0;
+
+        for (int i = 0; i < 1_000_000; i++) {   // repeat to get C2 to kick in
+            // Copy only latin1 chars from UTF-16 converted prefix (3 chars -> 3 bytes)
+            int intResult = (int) mCompressChars.invoke(null, chars, 0, bytes, 0, chars.length);
+            if (intResult == 0) {
+                if (printWarningCount == 0) {
+                    printWarningCount = 1;
+                    System.out.println("StringUTF16.compress returned 0, may not be intrinsic");
```

Comment:
It seems to me that the Java code for `StringUTF16.compress` also returns the index of the non-latin1 char, so I'm not sure I understand this. Just caution?

Reply:
There are separate implementations of the intrinsic for each platform.
This test would help identify the platforms on which the intrinsic had not been updated to return the index instead of zero.

Reply:
Ah OK.
So maybe the message should read `"intrinsic for StringUTF16.compress returned 0, may not have been updated"`